published/external/xdpapi.h

```diff
@@ -156,7 +228,269 @@ XdpUnloadApi(
     FreeLibrary(XdpHandle);
 }
 
-#endif // !defined(_KERNEL_MODE)
+#endif // defined(_KERNEL_MODE)
+
+#if defined(_KERNEL_MODE)
+
+#include <netioddk.h>
+#include <xdp/guid.h>
+
+#define POOLTAG_NMR_CLIENT      'cNdX'   // XdNc
+
+typedef struct _XDP_LOAD_CONTEXT *XDP_LOAD_API_CONTEXT;
+
+typedef
+VOID
+XDP_API_ATTACH_FN(
+    _In_ VOID *ClientContext
+    );
+
+typedef
+VOID
+XDP_API_DETACH_FN(
+    _In_ VOID *ClientContext
+    );
+
+typedef struct _XDP_API_CLIENT {
+    ERESOURCE Resource;
```

Comment:
until now, xdp has always used push locks for locking. can we use push locks here, too?

Reply:
Is there any reference implementation of the push locks?

Reply:
MsQuic has an example of usage [here](https://github.com/microsoft/msquic/blob/main/src/inc/quic_platform_winkernel.h#L287) for `EX_PUSH_LOCK`. Replace `ERESOURCE` usage with `EX_PUSH_LOCK`.

Reply:
What does this mean? There is strong reason to use RESOURCE instead of push lock?

https://github.com/microsoft/xdp-for-windows/blob/d5c8b5ae42d8e73dcd0bcc2f9467b3ac3154f64c/published/external/xdpapi.h#L405-L408

Reply:
The biggest reason is that pushlocks can't fail to initialize, which simplifies error handling, and also in general we should use the fewest possible patterns/data types. Since we've already used pushlocks for passive-level locking throughout XDP, we should continue to use them.

All that being said, pushlocks have some other nice properties, such as disallowing recursive exclusive acquisition (which tends to be a dangerous feature) and are smaller.